internal/fs/buffer/inmemory_write_buffer.go

```diff
@@ -15,33 +15,165 @@
 package buffer
 
 import (
-	"bytes"
+	"fmt"
+
+	"github.com/googlecloudplatform/gcsfuse/internal/logger"
 )
 
+type offset struct {
+	// [ start, end) is the range of file offset for which the
+	// data is currently stored in the buffer.
+	start, end int64
+}
+
+func (o *offset) advanceBy(value int64) {
+	o.start += value
+	o.end += value
+}
+
+type inMemoryBuffer struct {
+	buffer []byte
+	offset offset
+}
+
+// InMemoryWriteBuffer implements WriteBuffer and is stored in memory.
+// Refer to buffer.WriteBuffer for sample code usage.
 type InMemoryWriteBuffer struct {
-	// Holds the data to be written to GCS. At any time, Buffer holds the data for
-	// 2 GCS write calls.
-	buffer *bytes.Buffer
+	// Holds the incoming data from kernel write calls, currently being buffered
+	// for upload in the future.
+	current inMemoryBuffer
+
+	// Holds the data that is currently being uploaded asynchronously to GCS. Data
+	// in this buffer will persist until the entire data is successfully uploaded.
+	flushed inMemoryBuffer
+
+	// Holds the last offset to which data is written.
+	fileSize int64
+
+	// Holds the buffer's size (in bytes) to be allocated to currentBuffer and flushedBuffer.
```

Comment:
Nit: comment needs an update. 
```suggestion
	// Holds the buffer's size (in bytes) to be allocated to current and flushed.
```